utils/clickhouse.hh

```diff
@@ -0,0 +1,102 @@
+/**
+ *		Tempesta FW
+ *
+ * Copyright (C) 2024 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#ifndef __TFW_CLICKHOUSE_HH__
+#define __TFW_CLICKHOUSE_HH__
+
+#include <iostream>
+
+#include <clickhouse/base/socket.h>
+#include <clickhouse/client.h>
+
+/**
+ * Class for sending records to a Clickhouse database.
+ *
+ * Constructor:
+ *    @TfwClickhouse - Initializes the Clickhouse connection, and create a data
+ *        block, with a provided callback.
+ *
+ * Destructor:
+ *    @~TfwClickhouse - Delete Block object in block_.
+ *
+ * Other public methods:
+ *    @get_block - Returns a pointer to the data block for the specified CPU core.
+ *    @commit - Commits the data in the block to the Clickhouse database if the
+ *        elapsed time since the last insertion exceeds a predefined threshold
+ *        or the block’s row count exceeds a maximum event threshold. After
+ *        committing, the block is deleted, a new block is created via
+ *        block_callback and last_time is updated.
+ *
+ * Private Members:
+ *    @client_ - Clickhouse Client instance for sending data to the database.
+ *    @block_ - Block instance holding data records to be inserted.
+ *    @last_time_ - The last timestamp when data was sent.
+ *    @block_callback_ - Callback function that creates a new Block for data
+ *        storage.
+ *    @table_name_ - Name of the Clickhouse table where data is inserted.
+ */
+class TfwClickhouse {
+public:
+	TfwClickhouse(std::string host, std::string table_name,
+		      std::string user, std::string password,
+		      std::shared_ptr<clickhouse::Block> (*cb)());
+	TfwClickhouse(const TfwClickhouse &) = delete;
+	TfwClickhouse &operator=(const TfwClickhouse &) = delete;
+
+	std::shared_ptr<clickhouse::Block> get_block();
+	void commit();
+
```

Comment:
Don't we need a destructor flushing current block?

Reply:
You are right, it make sense to send accumulated data.
Done.